libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs

```diff
@@ -495,154 +59,234 @@ stsProperty ::
-  (PropEnv era -> GenShrink st) ->
-  (PropEnv era -> st -> GenShrink sig) ->
-  (PropEnv era -> env -> st -> sig -> st -> Property) ->
-  Property
-stsProperty sub genEnv genSt genSig prop =
-  lookupTerm sub network $ \peNetwork ->
-    lookupTerm sub currentEpoch $ \peEpochNo ->
-      lookupTerm sub currentSlot $ \peSlotNo ->
-        forAllBlind (pParamsGen sub) $ \(unPParams -> pePParams) ->
-          let pEnv = PropEnv {..}
-           in uncurry forAllShrinkBlind (genEnv pEnv) $ \env ->
-                counterexample (show $ ppString "env = " <> prettyA env) $
-                  uncurry forAllShrinkBlind (genSt pEnv) $ \st ->
-                    counterexample (show $ ppString "st = " <> prettyA st) $
-                      uncurry forAllShrinkBlind (genSig pEnv st) $ \sig ->
-                        counterexample (show $ ppString "sig = " <> prettyA sig) $
-                          runShelleyBase $ do
-                            res <- applySTS @(EraRule r era) $ TRC (env, st, sig)
-                            pure $ case res of
-                              Left pfailures -> counterexample (show $ ppList prettyA pfailures) $ property False
-                              Right st' ->
-                                counterexample (show $ ppString "st' = " <> prettyA st') $
-                                  prop pEnv env st sig st'
-
-checkConstraints ::
-  era ~ ConwayEra StandardCrypto =>
-  Proof era ->
-  Subst (ConwayEra StandardCrypto) ->
-  (Proof era -> [Pred era]) ->
-  RootTarget era s t ->
-  s ->
+  Spec fn env ->
+  (env -> Spec fn st) ->
+  (env -> st -> Spec fn sig) ->
+  (env -> st -> sig -> st -> p) ->
   Property
-checkConstraints proof sub preds target val =
-  let cs = applySubst sub $ preds proof
-      env = saturateEnv (getTarget val target emptyEnv) cs
-      check p = counterexample ("Failed: " ++ showPred p) $ ensureTyped (runPred env p) id
-   in counterexample ("-- Solution --\n" ++ showEnv env) $
-        conjoin $
-          map check cs
+stsPropertyV2 specEnv specState specSig prop =
+  uncurry forAllShrinkBlind (genShrinkFromSpec specEnv) $ \env ->
+    counterexample (show $ ppString "env = " <> prettyA env) $
+      uncurry forAllShrinkBlind (genShrinkFromSpec $ specState env) $ \st ->
+        counterexample (show $ ppString "st = " <> prettyA st) $
+          uncurry forAllShrinkBlind (genShrinkFromSpec $ specSig env st) $ \sig ->
+            counterexample (show $ ppString "sig = " <> prettyA sig) $
+              runShelleyBase $ do
+                res <- applySTS @(EraRule r era) $ TRC (env, st, sig)
+                pure $ case res of
+                  Left pfailures -> counterexample (show $ ppList prettyA pfailures) $ property False
+                  Right st' ->
+                    counterexample (show $ ppString "st' = " <> prettyA st') $
+                      conformsToSpec @fn st' (specState env)
+                        .&&. prop env st sig st'
```

Comment:
Continuation passing style makes this look really ugly with all the staircasing. Would it be possible to do this with monadic binds instead?

Reply:
Yes. I agree. Fourmolololol made me do it! I don't know if there is a monadic interface for QuickCheck that will let me write this conveniently and have it mean the same thing, but I'll give it a try.

@lehins case in point ;)

Reply:
For example, if one were to switch to `PropertyM` and use `forAllM` we would lose shrinking completely. (For whatever reason, `forAllM` quantified values don't shrink).

Reply:
Without fourmolu you could write it like this:
```
  -- Generate the environment
  uncurry forAllShrinkBlind (genShrinkFromSpec specEnv)          $ \env ->
  counterexample (show $ ppString "env = " <> prettyA env) $

  -- Generate the state
  uncurry forAllShrinkBlind (genShrinkFromSpec $ specState env)  $ \st ->
  counterexample (show $ ppString "st = " <> prettyA st) $

  -- Generate the signal
  uncurry forAllShrinkBlind (genShrinkFromSpec $ specSig env st) $ \sig ->
  counterexample (show $ ppString "sig = " <> prettyA sig) $

  -- Run the rule and check that the result conforms to the spec
  runShelleyBase $ do
    res <- applySTS @(EraRule r era) $ TRC (env, st, sig)
    pure $ case res of
      Left pfailures -> counterexample (show $ ppList prettyA pfailures) $ property False
      Right st' ->
        counterexample (show $ ppString "st' = " <> prettyA st') $
          conformsToSpec @fn st' (specState env)
            .&&. prop env st sig st'
```

However, here are a few things you _can't_ do:
1. Generate all the things at the same time (because if generation fails for one thing you wouldn't get a QuickCheck `counterexample` for the earlier things)
2. Replace the multiple calls to `counterexample` with one call to `counterexample` (for the same reason, that means that if generation fails for the signal you don't get to see what the environment and the state were).

Reply:
The interface to QuickCheck really does force you to write your code like this, and fourmolu then makes it unreadable...

Reply:
> fourmolu then makes it unreadable...

On the contrary, this is horrendous and unreadable: 

```
  -- Generate the environment
  uncurry forAllShrinkBlind (genShrinkFromSpec specEnv)          $ \env ->
  counterexample (show $ ppString "env = " <> prettyA env) $

  -- Generate the state
  uncurry forAllShrinkBlind (genShrinkFromSpec $ specState env)  $ \st ->
  counterexample (show $ ppString "st = " <> prettyA st) $

  -- Generate the signal
  uncurry forAllShrinkBlind (genShrinkFromSpec $ specSig env st) $ \sig ->
  counterexample (show $ ppString "sig = " <> prettyA sig) $
```

It is totally unclear where the scope of one functions begins and the other one ends. I would never allow code like this to be merged, even without fourmolu.

So, let's forget about the style and focus on semantics! That is precisely what the automatic formatter is for, so people can put their opinions about the style aside and focus on what is important.

Question was: can we change from the CPS style to a more readable one?
Seems like the answer is: no.

So. I am marking this issue as resolved.

Reply:
Let's agree to disagree. But if line breaks and lambda bother you, do line breaks in do statements also make the code horrendous?

Reply:
In my opinion current code looks totally fine. I am not bothered neither by lambdas, nor by stair stepping. I personally was more curious if it was possible to make it a little simpler, but if that is too much work or is not possible then there is nothing need to be done here.